server/service/integration_mdm_test.go

```diff
@@ -317,6 +529,60 @@ func (s *integrationMDMTestSuite) TestAppleMDMCSRRequest() {
 	require.Contains(t, string(reqCSRResp.SCEPKey), "-----BEGIN RSA PRIVATE KEY-----\n")
 }
 
+func (s *integrationMDMTestSuite) TestMDMAppleUnenroll() {
+	t := s.T()
+	// enroll into mdm
+	d := newMDMEnrolledDevice(s)
+
+	// set an enroll secret
+	var applyResp applyEnrollSecretSpecResponse
+	s.DoJSON("POST", "/api/latest/fleet/spec/enroll_secret", applyEnrollSecretSpecRequest{
+		Spec: &fleet.EnrollSecretSpec{
+			Secrets: []*fleet.EnrollSecret{{Secret: t.Name()}},
+		},
+	}, http.StatusOK, &applyResp)
+
+	// simulate a matching host enrolling via osquery
+	j, err := json.Marshal(&enrollAgentRequest{
+		EnrollSecret:   t.Name(),
+		HostIdentifier: d.uuid,
+	})
+	require.NoError(t, err)
+	var enrollResp enrollAgentResponse
+	hres := s.DoRawNoAuth("POST", "/api/osquery/enroll", j, http.StatusOK)
+	defer hres.Body.Close()
+	require.NoError(t, json.NewDecoder(hres.Body).Decode(&enrollResp))
+	require.NotEmpty(t, enrollResp.NodeKey)
+
+	listHostsRes := listHostsResponse{}
+	s.DoJSON("GET", "/api/latest/fleet/hosts", nil, http.StatusOK, &listHostsRes)
+	require.Len(t, listHostsRes.Hosts, 1)
+	h := listHostsRes.Hosts[0]
+
+	// try to unenroll the host, fails since the host doesn't respond
+	s.Do("PATCH", fmt.Sprintf("/api/latest/fleet/mdm/hosts/%d/unenroll", h.ID), nil, http.StatusGatewayTimeout)
+
+	// we're going to modify this mock, make sure we restore its default
+	originalPushMock := s.pushProvider.PushFunc
+	defer func() { s.pushProvider.PushFunc = originalPushMock }()
+
+	// TODO: this is not working as expected, we're still waiting on the
+	// device to unenroll even if we weren't able to send a push
+	// notification, we should return an error instead.
+	//
+	// the APNs service returns an error
+	// s.pushProvider.PushFunc = mockFailedPush
+	// s.Do("PATCH", fmt.Sprintf("/api/latest/fleet/mdm/hosts/%d/unenroll", h.ID), nil, http.StatusOK)
```

Comment:
@gillespi314 @mna I think we should move away from [rawCommandEnqueue](https://github.com/fleetdm/fleet/blob/2b3ddaa4b3662f0c04e3db7fc0859a3565bb1074/server/service/apple_mdm.go#L540) to enqueue Fleet internal commands and have a service method instead, we will also want to start keeping some kind of record in the DB each time a command is issues by fleet so we can track its status.

Do you have any thoughts about this?

Reply:
Yeah that makes sense to me. :+1:

Reply:
Totally agree